### PR TITLE
DEL-273:Monitor SSO

### PR DESCRIPTION
**Pull request by STEPHINRACHEL**
**What**
The initial draft for monitoring SSO 

**To Verify**
1. Checkout branch 
2. `docker build -t quay.io/replace-with-your-repo/workload-app .`
3. `docker push quay.io/replace-with-your-repo/workload-app`
4. Update the image reference in the `deploy/template.yaml `
5. Login to your Openshift  cluster using `oc login ..`
6. Run `make deploy`
7. Check the pod log in the workload-web-app namespace, you should see _Login 
succeeded_ message in the log

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -51,8 +52,34 @@ func startAMQChecks() {
...
+func startSSOChecks() {
+	url := "https://keycloak-edge-redhat-rhmi-rhsso.apps.sso-128.y3z4.s1.devshift.org"
+	user := "admin"
+	pwd := "NH0trX4G93snKw=="
+	realm := "master"
+	if url != "" && user != "" && pwd != "" && realm != "" {
+		log.WithFields(log.Fields{
+			"serverURL": url,
+			"user":      user,
+			"password":  pwd,
```
Comment (wei-lee) [inline main.go]: be careful not to log sensitive data in logs.

File: sso.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	} else {
+		ssoLoginGauge.Set(1)
+		log.Info("Login succeeded!!")
+	}
+}
+
+func (s *SSOChecks) runForever() {
+	for {
+		s.run()
+		time.Sleep(s.interval)
```
Comment (wei-lee) [inline sso.go:48]: I can't remember where I saw it but I think SSO has rate limit in place for login requests. We might need to take that into consideration and have a minimum value set for the interval.

File: main.go (hunk 1 above)
Comment (STEPHINRACHEL) [reply main.go]: Sure. Will remove the sensitive data from logs

File: sso.go (hunk 2 above)
Comment (STEPHINRACHEL) [reply sso.go:49]: @wei-lee I could see the global rate limit for user/password authentication as 10 attempts/minute   [Rate limits on user/password auth](https://auth0.com/docs/connections/database/rate-limits) so we can keep the interval to a minimum of  6s or 7s. Does that sound good?

File: main.go [hunk 3]
```diff
@@ -53,23 +56,21 @@ func startAMQChecks() {
...
 			"realmName": realm,
-		}).Info("Start SSO Checks!!")
+		}).Info("Start SSO Checks")
 		c := &SSOChecks{
 			serverURL: url,
 			user:      user,
 			password:  pwd,
 			realmName: realm,
-			interval:  1 * time.Second,
+			interval:  7 * time.Second,
```
Comment (STEPHINRACHEL) [inline main.go:73]: Interval is set to 7s for avoiding the risk of getting the login action blocked by the rate limit of SSO.

File: deploy/template.yaml [hunk 4]
```diff
@@ -60,4 +66,10 @@ parameters:
...
   - name: AMQ_QUEUE_NAME
-    description: The queue name to use for the AMQ address
\ No newline at end of file
+    description: The queue name to use for the AMQ address
+  - name: SERVER_URL
+    description: The RHSSO Server URL
+  - name: USER
+    description: The username for RHSSO login
+  - name: PWD
+    description: The Password for RHSSO login
```
Comment (mikenairn) [inline deploy/template.yaml:81]: Should we add RHSSO_ to all of these to make it more obvious what they are for? (RHSSO_SERVER_URL, RHSSO_USER, RHSSO_PWD etc...)

Comment (STEPHINRACHEL) [reply deploy/template.yaml:81]: Sounds good

File: deploy/deploy.sh [hunk 5]
```diff
@@ -39,14 +39,25 @@ retry 20 5 oc get address/workload-app.queue-requests -n $NS -o 'jsonpath={.stat
 AMQ_ADDRESS="amqps://$(oc get addressspace/workload-app -n $NS -o 'jsonpath={.status.endpointStatuses[?(@.name=="messaging")].serviceHost}')"
 AMQ_QUEUE="/$(oc get address/workload-app.queue-requests -n $NS -o 'jsonpath={.spec.address}')"
 
+#SSO credentials
+RHSSO_SERVER_URL="https://$(oc get routes -n redhat-rhmi-rhsso keycloak-edge -o 'jsonpath={.spec.host}')"
```
Comment (b1zzu) [inline deploy/deploy.sh]: I thought we should test the user sso and not the cluster one `redhat-rhmi-user-sso`

Comment (mikenairn) [reply deploy/deploy.sh]: I cant remember what was decided here, or really ever talking about it tbh, is there any difference doing one over the other?

Comment (STEPHINRACHEL) [reply deploy/deploy.sh]: Hmm, Just noticed now in the task description, it should be user RHSSO. I was sticking to the RHSSO in the task title 😀. I am also not sure about any significant difference in doing one over the other

Comment (mikenairn) [reply deploy/deploy.sh]: OK, if we said user rhsso, i guess we better change it here.

File: deploy/template.yaml [hunk 6]
```diff
@@ -36,6 +36,12 @@ objects:
                   value: ${AMQ_QUEUE_NAME}
                 - name: ENVIRONMENT
                   value: production
+                - name: RHSSO_SERVER_URL
+                  value: ${RHSSO_SERVER_URL}
+                - name: RHSSO_USER
+                  value: ${RHSSO_USER}
+                - name: RHSSO_PWD
+                  value: ${RHSSO_PWD}
```
Comment (wei-lee) [inline deploy/template.yaml]: let's create a secret and store the password in the secret and then we can reference the secret from here.

File: deploy/deploy.sh (hunk 5 above)
Comment (b1zzu) [reply deploy/deploy.sh]: The user sso is the instance used by the customer to provide sso access to them for their users to their apps, therefore from an availability point of view is more critical that the internal sso that is used only to login to the RHMI consoles, that's why I remember we said that we were going to test the user sso

Comment (STEPHINRACHEL) [reply deploy/deploy.sh]: Good catch @b1zzu

**Comment by mikenairn** [comment]
Works as expected, metric during normal operation is 1:

```
$ oc expose svc/workload-web-app --path /metrics --name metrics
$ curl -s http://metrics-workload-web-app.apps.mnairn.l9z7.s1.devshift.org/metrics | grep ^sso
sso_login_success 1
```

When i disrupted the user sso pods i could see the following error message in the logs:

```
time="2020-06-03T14:43:14Z" level=warning msg="Login failed with error  :503 Service Unavailable"
```
And the metric exposed changes:

```
$ curl -s http://metrics-workload-web-app.apps.mnairn.l9z7.s1.devshift.org/metrics | grep ^sso
sso_login_success 0
```

/lgtm

**Pull request merged**